### PR TITLE
Feature/task 5087 add new attributes to add edit person place screen

### PR DESCRIPTION
**Pull request by KaiTeuber**
## Description
Add and edit contact diary entries (person or location) extended for multiple fields:
* name | description
* email address
* phone number

## Link to Jira
https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-5087

## Screenshots
Simulator Screen Shot - Clone 1 of iPhone 8 - 2021-02-16 at 18 09 23
Simulator Screen Shot - Clone 1 of iPhone 8 - 2021-02-16 at 18 09 45
Simulator Screen Shot - Clone 1 of iPhone 8 - 2021-02-16 at 18 10 14
Simulator Screen Shot - Clone 1 of iPhone 8 - 2021-02-16 at 18 10 20
feat
Simulator Screen Shot - iPhone 8 - 2021-02-16 at 18 10 54
Simulator Screen Shot - iPhone 8 - 2021-02-16 at 18 10 57
Simulator Screen Shot - iPhone 8 - 2021-02-16 at 18 11 19fea

### CODE REVIEW

**Review by tcfos — APPROVED** [review]

**Review by janetback — APPROVED** [review]

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/DiaryAddAndEditEntryViewController.swift [hunk 1]
```diff
@@ -73,34 +78,41 @@ class DiaryAddAndEditEntryViewController: UIViewController, UITextFieldDelegate,
...
+		case .default, .done, .send:
+			if !viewModel.entryModel.isEmpty {
+				viewModel.save()
+			}
+			textfieldsManager.resignFirstResponder()
+			dismiss()
+		case .next, .continue:
+			textfieldsManager.nextFirstResponder()
+		default:
+			Log.debug("unsupport return key type")
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/DiaryAddAndEditEntryViewController.swift:91]: Should we call textfieldsManager.nextFirstResponder() here?

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/__test__/TextFieldsManagerTests.swift [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+		// GIVEN
+		let manager = TextFieldsManager()
+		let textField1 = UITextField()
+		let textField2 = UITextField()
+
+		manager.appendTextField(textfiledWithKayPath: (textField1, \DiaryAddAndEditEntryModel.name))
+		manager.appendTextField(textfiledWithKayPath: (textField2, \DiaryAddAndEditEntryModel.phoneNumber))
+
+		// WHEN
+		manager.nextFirstResponder()
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/__test__/TextFieldsManagerTests.swift:63]: Maybe we should check XCTAssertTrue(firstResponder1) somewhere here?

**Review by ArturFriesen — APPROVED** [review]

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/DiaryAddAndEditEntryViewController.swift (hunk 1 above)
Comment (KaiTeuber) [reply src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/DiaryAddAndEditEntryViewController.swift:91]: no, I think the default behavior should be to enable next first responder only on .next or .continue. But we are open here to change things later if needed

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/DiaryAddAndEditEntryModel.swift [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+		case .contactPerson:
+			self.namePlaceholder = AppStrings.ContactDiary.AddEditEntry.person.placeholders.name
+			self.phonenumberPlaceholder = AppStrings.ContactDiary.AddEditEntry.person.placeholders.phonenumber
+			self.emailAddressPlaceholder = AppStrings.ContactDiary.AddEditEntry.person.placeholders.email
+
+		case .location:
+			self.namePlaceholder = AppStrings.ContactDiary.AddEditEntry.location.placeholders.name
+			self.phonenumberPlaceholder = AppStrings.ContactDiary.AddEditEntry.location.placeholders.phonenumber
+			self.emailAddressPlaceholder = AppStrings.ContactDiary.AddEditEntry.location.placeholders.email
+		}
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/DiaryAddAndEditEntryModel.swift:22]: If this was extracted into something like `private func setPlaceholders(for entryType: EntryType)` you could reuse it in the other inits. Just a suggestion :)

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/DiaryAddAndEditEntryViewController.swift [hunk 4]
```diff
@@ -73,34 +78,41 @@ class DiaryAddAndEditEntryViewController: UIViewController, UITextFieldDelegate,
...
 	func wasAttemptedToBeDismissed() {
-		entryTextField.resignFirstResponder()
+		textfieldsManager.resignFirstResponder()
 		dismiss()
 	}
 
 	// MARK: - Private
 
 	private let viewModel: DiaryAddAndEditEntryViewModel
+	private let textfieldsManager: TextFieldsManager
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/DiaryAddAndEditEntryViewController.swift:115]:
```suggestion
	private let textFieldsManager: TextFieldsManager
```

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/DiaryAddAndEditEntryModel.swift [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+		self.name = person.name
+		self.phoneNumber = person.phoneNumber
+		self.emailAddress = person.emailAddress
+	}
+
+	// MARK: - Internal
+
+	let namePlaceholder: String
+	let phonenumberPlaceholder: String
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/DiaryAddAndEditEntryModel.swift:52]:
```suggestion
	let phoneNumberPlaceholder: String
```

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/DiaryAddAndEditEntryModel.swift [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+		}
+
+		self.name = ""
+		self.phoneNumber = ""
+		self.emailAddress = ""
+	}
+
+	init(_ location: DiaryLocation) {
+		self.namePlaceholder = AppStrings.ContactDiary.AddEditEntry.location.placeholders.name
+		self.phonenumberPlaceholder = AppStrings.ContactDiary.AddEditEntry.location.placeholders.phonenumber
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/DiaryAddAndEditEntryModel.swift:31]:
```suggestion
		self.phonenumberPlaceholder = AppStrings.ContactDiary.AddEditEntry.location.placeholders.phoneNumber
```

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/DiaryAddAndEditEntryModel.swift [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+		self.emailAddressPlaceholder = AppStrings.ContactDiary.AddEditEntry.location.placeholders.email
+
+		self.name = location.name
+		self.phoneNumber = location.phoneNumber
+		self.emailAddress = location.emailAddress
+	}
+
+	init(_ person: DiaryContactPerson) {
+		self.namePlaceholder = AppStrings.ContactDiary.AddEditEntry.person.placeholders.name
+		self.phonenumberPlaceholder = AppStrings.ContactDiary.AddEditEntry.person.placeholders.phonenumber
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/DiaryAddAndEditEntryModel.swift:41]:
```suggestion
		self.phonenumberPlaceholder = AppStrings.ContactDiary.AddEditEntry.person.placeholders.phoneNumber
```

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/DiaryAddAndEditEntryViewModel.swift [hunk 8]
```diff
@@ -18,14 +18,16 @@ class DiaryAddAndEditEntryViewModel {
...
 		case .edit(let entry):
 			switch entry {
 			case .location(let location):
-				self.textInput = location.name
+				self.entryModel = DiaryAddAndEditEntryModel(location)
+
 			case .contactPerson(let person):
-				self.textInput = person.name
+				self.entryModel = DiaryAddAndEditEntryModel(person)
 			}
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/DiaryAddAndEditEntryViewModel.swift:31]: Could be shortened to `self.entryModel = DiaryAddAndEditEntryModel(entry.type)`

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/TextFieldManager.swift [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+	}
+
+	// MARK: - Internal
+
+	// helper to get all textfiled without attached KeyPath
+	var textFiledsOnly: [UITextField] {
+		textFieldsWithKeyPaths.map { textFiled, _ -> UITextField in
+			return textFiled
+		}
+	}
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/TextFieldManager.swift:24]:
```suggestion
	var textFieldsOnly: [UITextField] {
		textFieldsWithKeyPaths.map { textField, _ -> UITextField in
			return textField
		}
	}
```

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/DiaryAddAndEditEntryViewModel.swift [hunk 10]
```diff
@@ -86,13 +86,16 @@ class DiaryAddAndEditEntryViewModel {
...
-		case .add(_, let entryType):
-			return placeholderText(from: entryType)
-		case .edit(let entry):
-			return placeholderText(from: entry.type)
-		}
+	var namePlaceholer: String {
+		return entryModel.namePlaceholder
+	}
+
+	var phonenumberPlaceholder: String {
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/DiaryAddAndEditEntryViewModel.swift:93]:
```suggestion
	var phoneNumberPlaceholder: String {
```

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/__test__/TextFieldsManagerTests.swift (hunk 2 above)
Comment (KaiTeuber) [reply src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/__test__/TextFieldsManagerTests.swift:73]: I added a new test to check that as well

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/DiaryAddAndEditEntryViewModel.swift [hunk 11]
```diff
@@ -86,13 +86,16 @@ class DiaryAddAndEditEntryViewModel {
...
+		return entryModel.phonenumberPlaceholder
+	}
+
+	var emailAddressPlacehodler: String {
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/DiaryAddAndEditEntryViewModel.swift:97]:
```suggestion
	var emailAddressPlaceholder: String {
```

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/TextFieldManager.swift [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+
+	// get the associated WritableKeyPath for a textfield - if known
+	func keyPath(for searchTextField: UITextField) -> WritableKeyPath<DiaryAddAndEditEntryModel, String>? {
+		textFieldsWithKeyPaths.first { textField, _ -> Bool in
+			searchTextField == textField
+		}?.1
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/TextFieldManager.swift:30]: Not sure, could this `1` be named somehow?

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/TextFieldManager.swift [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+	// will become the new first responder
+	func nextFirstResponder() {
+		guard let currentFirstResponder = firstResponder,
+			  let index = textFiledsOnly.firstIndex(of: currentFirstResponder) else {
+			textFiledsOnly.first?.becomeFirstResponder()
+			return
+		}
+		let nextIndex = index + 1
+		guard textFiledsOnly.indices.contains(nextIndex) else {
+			Log.debug("there is no next textfiled to enable as first responder", log: .default)
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/TextFieldManager.swift:49]:
```suggestion
			Log.debug("there is no next text field to enable as first responder", log: .default)
```

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/TextFieldManager.swift [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+			return
+		}
+		textFiledsOnly[nextIndex].becomeFirstResponder()
+	}
+
+	// method to add a new textfiled to the known array of TextFields with attatched KeyPath
+	func appendTextField(textfiledWithKayPath: (UITextField, WritableKeyPath<DiaryAddAndEditEntryModel, String>)) {
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/TextFieldManager.swift:56]:
```suggestion
	func appendTextField(textFieldWithKeyPath: (UITextField, WritableKeyPath<DiaryAddAndEditEntryModel, String>)) {
```

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/TextFieldManager.swift [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+		guard !textFiledsOnly.contains(textfiledWithKayPath.0) else {
+			Log.debug("Textfiled already known - stop here", log: .default)
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/TextFieldManager.swift:58]:
```suggestion
			Log.debug("Text field already known - stop here", log: .default)
```

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/TextFieldManager.swift [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+		textFieldsWithKeyPaths.append(textfiledWithKayPath)
+	}
+
+	// MARK: - Private
+
+	private var textFieldsWithKeyPaths: [(UITextField, WritableKeyPath<DiaryAddAndEditEntryModel, String>)]
+
+	/// look for the current first responder and return if available
+	private var firstResponder: UITextField? {
+		return textFiledsOnly.first { textfield -> Bool in
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/TextFieldManager.swift:70]:
```suggestion
		return textFiledsOnly.first { textField -> Bool in
```

**Review by nickguendling — APPROVED** [review]
Looking good to me 👍 Just found a couple of typos and I think two other things that could be improved, but just as a suggestion.

File: src/xcode/ENA/ENA/Resources/Localization/en.lproj/Localizable.strings [hunk 17]
```diff
@@ -1466,12 +1466,12 @@
 /* - Location */
 "ContactDiary_AddEditEntry_LocationTitle" = "Place";
 
-"ContactDiary_AddEditEntry_LocationPlaceholder" = "Place";
+"ContactDiary_AddEditEntry_LocationPlaceholder_Name" = "Place";
```
[the remark this answers is not in the document]
Comment (KaiTeuber) [reply src/xcode/ENA/ENA/Resources/Localization/en.lproj/Localizable.strings:1469]: I had a shot chat about that with Janet, she approved it

File: src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/TextFieldManager.swift (hunk 12 above)
Comment (KaiTeuber) [reply src/xcode/ENA/ENA/Source/Scenes/ContactDiary/AddAndEditEntry/TextFieldManager.swift:30]: done

**Pull request merged**